pilot/pkg/security/model/authentication.go

```diff
@@ -205,8 +209,38 @@ func ConstructValidationContext(rootCAFilePath string, subjectAltNames []string)
 	return ret
 }
 
+// ConstructSANWithTrustDomain constructs MatchSubjectAltNames to be used in DefaultValidationContext by merging
+// Exact SAN matcher user has supplied at Gateway and Prefix SAN matcher for trust domain validation
+func ConstructSANWithTrustDomain(in []string, trustDomainAliases []string) []*matcher.StringMatcher {
+	//  initially contains only usersupplied SAN
+	userSupplied := util.StringToExactMatch(in)
+	// Early exit if Trust Domain not to be verified
+	if features.SkipValidateTrustDomain.Get() {
+		return userSupplied
+	}
+	// Add Trust Domain SAN Prefix matching if feature's enabled
+	trustDomain := spiffe.URIPrefix + spiffe.GetTrustDomain()
+
+	tdMatcher := make([]*matcher.StringMatcher, 0, 1+len(trustDomainAliases))
+	// append TD from istiod
+	tdMatcher = append(tdMatcher, &matcher.StringMatcher{
+		MatchPattern: &matcher.StringMatcher_Prefix{Prefix: trustDomain},
+	})
+	// append TD aliases from meshconfig
+	for _, a := range trustDomainAliases {
+		tdMatcher = append(tdMatcher, &matcher.StringMatcher{
+			MatchPattern: &matcher.StringMatcher_Prefix{Prefix: spiffe.URIPrefix + a},
+		})
+	}
+
+	res := append(userSupplied, tdMatcher...)
+
+	return res
+}
+
 // ApplyToCommonTLSContext completes the commonTlsContext for `ISTIO_MUTUAL` TLS mode
-func ApplyToCommonTLSContext(tlsContext *tls.CommonTlsContext, metadata *model.NodeMetadata, sdsPath string, subjectAltNames []string) {
+func ApplyToCommonTLSContext(tlsContext *tls.CommonTlsContext, metadata *model.NodeMetadata,
```

Comment:
instead of adding `isGateway bool`, can we just check if `trustDomainAliases` is not empty, then append it to the matchSAN?

Reply:
That doesn't work for this use case, because `trustDomainAliases` don't contain local trust domain but local trust domain always exists.

Reply:
you can just append the `cluster.local` to the array when the feature flag is true. The ApplyToCommonTLSContext should be agnostic about where it is being called, the caller decide what to pass based on where it's called, and the function itself has a consistent behavior just based on its input (and less is better).